duckduckgo/BrowserServicesKit
# Allow automated fetching of synced bookmarks' favicons

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1205949780297088/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2163
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1857
What kind of version bump will this require?: Major

Tech Design URL: https://app.asana.com/0/481882893211075/1204986998781220/f

**Description**:
Add BookmarksFaviconFetcher that is used to fetch favicons for bookmarks received by Sync.
Fetcher is opt-in, controlled by a setting inside Sync settings (with an additional in-context onboarding
popup presented from client apps). Fetcher uses LinkPresentation framework to obtain a favicon
for a given domain, and in case of failure it falls back to checking hardcoded favicon URLs.
Fetcher keeps a state internally, by saving list of bookmarks IDs that need processing to a file on disk.
Fetcher plugs into clients' implementation of favicon storage by exposing FaviconStoring protocol.
Fetcher performs fetching on a serial operation queue. Each fetcher invocation cancels previously scheduled
operation and schedules a new one. Updating fetcher state is also scheduled on the operation queue -
state updates don't support cancelling and always finish before next operation is started.



**Steps to test this PR**:
1. Run the macOS app from Xcode, enable Sync logging in Debug -> Logging -> Sync.
2. Import some bookmarks (you can use a HTML file provided in the [test builds task](https://app.asana.com/0/0/1205845802272364/f) - it adds a bunch of popular websites as favorites).
3. Start sync, run the iOS app and connect to the Sync account – enable sharing favorites so that iOS gets the same set as macOS.

On iOS:
1. On iOS, go back from Sync settings to the home screen. Drag the favorites view or open a new tab. Verify that you're prompted to enable favicons fetching.
2. Enable favicons fetching and observe that favicons are being fetched (you need to drag the view to trigger reloading, or open new tab page to reload the full view. Some of the favicons in the test set won't be fetched because bookmarks URLs are not found (e.g. https://www.dropboxusercontent.com)
3. Kill the app while fetching and restart. Verify that fetching is resumed.
4. Disable Sync. Verify that fetching is interrupted.
5. Re-enable Sync. Verify that fetching is disabled.
6. Go to Debug menu -> Sync -> Reset favicons fetcher onboarding.
7. Go to bookmarks view and scroll to a bookmark without favicon. Verify that you're prompted to enable fetching favicons.
8. Verify that after all favicons are fetched, fetcher doesn't try to fetch any more favicons after every sync operation (the fetching operation should start and immediately finish).

On macOS:
1. After enabling Sync on iOS, go back to new tab page. Verify that you're prompted to enable favicons fetching. Note that the popup UI is not final.
2. Enable favicons fetching and observe that favicons are being fetched (you need to drag the view or switch tabs back and forth to trigger reloading).
3. Kill the app while fetching and restart. Verify that fetching is resumed.
4. Disable Sync. Verify that fetching is interrupted.
5. Re-enable Sync. Verify that fetching is disabled.
6. Go to Debug menu -> Sync -> Reset favicons fetcher onboarding.
7. Open Bookmarks Management view. Verify that you're prompted to enable fetching favicons.
8. Verify that after all favicons are fetched, fetcher doesn't try to fetch any more favicons after every sync operation (the fetching operation should start and immediately finish).



**OS Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
